DOC-13291: Sync Gateway 3.3 Docs (#877)
couchbaselabs/docs-sync-gateway: modules/ROOT/pages/index-partitions.adoc

= Partitioned Indexes
ifdef::show_edition[:page-edition: {release}]
:page-status: Sync Gateway 3.3
:page-role:
:Description: Partitioning a large index across multiple nodes

include::partial$_set_page_context.adoc[]

:topic-group: Deploy
:param-related: {deployment--xref}  |  {rest-api-access--xref}  |  {configuration-properties-legacy--xref}
:param-abstract: Explains how to use partitioned indexes in Sync Gateway
include::partial$_show_page_header_block.adoc[]

== Overview

Partitioned indexes in Sync Gateway offer horizontal scalability for large deployments by sharding indexes across multiple nodes.
This can decrease performance, as each partition is queried separately and results are aggregated.

> Only `allDocs` and channels indexes can be partitioned.

Reviewer: ```suggestion
Only `allDocs` and `channels` indexes can be partitioned.
```

Either both indexes should be quoted or neither should be. I think it could go either way.

Author: Done

It increases resource usage (CPU and memory) on Query nodes, and may introduce performance degradation.
--

[#use-case]
=== Appropriate Use Case

* You are using MOI.
* Your current GSI index size exceeds the memory capacity of a single index node.
* You cannot further split your data (for example, by collections or database sharding) to reduce index size.

In this case, partitioning can help by horizontally scaling the index across multiple nodes to keep it in memory.

[#when-not-to-use]
=== When Not to Use Partitioned Indexes

* You're not using MOI; that is, your indexes are disk-based.
* Your dataset is small or moderate in size, even if you have high concurrency.
* You have a high number of connections or clients, but not many documents.
* Your index size could be reduced by splitting data into multiple collections, which is the preferred approach.

[#index-size]
=== What Impacts Index Size?

The size of Sync Gateway's indexes is correlated to:

* The number of documents in a collection.
* The number of channels each document is assigned to.

If index size is a concern, your first approach should be to:

* Split documents across multiple collections.

> * Review and minimize channel fan-out.

Reviewer: ```suggestion
```

Dropping the number of channels reduces the size of the `channels` indexes but it doesn't reduce the size of the `allDocs` index.

Author: Done

Only if these strategies are exhausted, and you're using MOI with oversized indexes, should partitioned indexes be considered.

[#migrate]
== Migrate to Partitioned Indexes

Before switching from non-partitioned to partitioned indexes, it's important to understand the operational trade-offs and cluster implications.
Here are two migration options, designed to accommodate different use cases and resource tolerances.

[#migrate-zero-downtime]
=== Option 1: Zero Downtime (Recommended for Production)

This option is ideal for production environments where uptime is critical.
It requires temporary over-provisioning of index nodes (2&times; size) while partitioned indexes are created.
The advantage is no downtime; the disadvantage is temporarily higher resource usage.

.Procedure
. Spin up a new Sync Gateway instance.

. Run xref:rest_api_admin_static.adoc#tag/Database-Management/operation/post_db-_index_init[`POST /{db}/_index_init`] on the new instance with `num_partitions` set to the required number of partitions.

. Wait for completion via xref:rest_api_admin_static.adoc#tag/Database-Management/operation/get_db-_index_init[`GET /{db}/_index_init`].

> . Run xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] to use new indexes.

Reviewer: ```suggestion
. Run xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] with `index.num_partitions` updated in the configuration to use new indexes.
```

Author: Done

WARNING: If you are using xref:configuration-overview.adoc#lbl-config-grp[configuration groups] running on a different index configuration, calling xref:rest_api_admin_static.adoc#tag/Server/operation/post__post_upgrade[`POST /_post_upgrade`] can delete indexes out from under a running database.
Please ensure that all configuration groups have consistent index configuration.
For details, see <<config-groups>>.

[#migrate-with-downtime]
=== Option 2: With Downtime

This option requires no extra resources, but involves planned downtime.
It involves lower intermittent hardware requirements.

.Procedure
. Take the database offline via xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] with `offline` set to `true`.

. Manually delete old Sync Gateway indexes.

. Bring the database online using xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/post_db-_config[`POST /{db}/_config`] with `index.num_partitions` set to the required number of partitions.

. Run integration and performance testing.

> WARNING: Manual index deletion can cause data loss or leave the system in an unstable state.

Author: @idulo @torcolvin see conversation at [#864](https://github.com/couchbase/docs-sync-gateway/pull/864/files/521c29e89bf6a5753499b2f00fb6e70027c464db#r2152378197)

Reverting requires downtime and manual deletion.
Make sure only advanced users (typically Architects or Admins) modify partitioning settings.

* Monitor memory and disk usage on index nodes during and after creation to catch unexpected bloat.

* Optionally set feature flags or config validation, if available, to restrict low-risk deployments from using this feature.

Changing the number of partitions "just to test it" may permanently increase index footprint, cause cluster performance issues, and trigger long rebalance operations.

[#config-groups]
== Configuration Group Considerations

If your deployment uses xref:configuration-overview.adoc#lbl-config-grp[Configuration Groups], ensure that:

* All nodes are updated with the new config before running xref:rest_api_admin_static.adoc#tag/Server/operation/post__post_upgrade[`POST /_post_upgrade`].
* Import and sync workloads are not unintentionally unbalanced across partitions.

Misalignment between config groups can lead to unexpected deletion of active indexes if xref:rest_api_admin_static.adoc#tag/Server/operation/post__post_upgrade[`POST /_post_upgrade`] is run.
This will result in:

> * Doing a CBS sequential scan on the Query nodes, which will seriously degrade performance of Query nodes.

Reviewer: ```suggestion
* If an index does not exist, Couchbase Server 7.6+ will fall back to performing a sequential scan on the Query nodes, which will seriously degrade performance of Query nodes.
```

If you are using Couchbase Server < 7.6, then the query will silently fail. I wish it wasn't this way, but there's no way to tell.

I'm fine if you don't want to make the change to this wording and it is too complicated.

Author: Done.

* Recreating indexes that were deleted by another config when a node is rebooted with a different `index.num_partitions`, even after xref:rest_api_admin_static.adoc#tag/Server/operation/post__post_upgrade[`POST /_post_upgrade`] is run.

[#supportability]
== Supportability

[#index-config]
=== Identify Index Configuration

To identify the current index configuration, you can:

* Use xref:rest_api_admin_static.adoc#tag/Database-Management/operation/get_db-_index_init[`GET /{db}/_index_init`] to verify the `settings.num_partitions` in use.

> * Use xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/get_db-_config[`GET /{db}/_config`] to check the active configuration.

Reviewer: ```suggestion
* Use xref:rest_api_admin_static.adoc#tag/Database-Configuration/operation/get_db-_config[`GET /{db}/_config`] to check the active configuration in `index.num_partitions`.

* Use xref:rest_api_admin_static.adoc#tag/Database-Management/operation/get_db-_index_init[`GET /{db}/_index_init`] to see if there is an index creation in progress.
```

Author: Done
